Add click-compatible pools command

hankehly opened the pull request:
Related: #22708

## Summary

This PR adds a click-compatible `pools` command group and sub-commands. The behavior is the exactly same as the non-click version.

- [x] add click command
- [x] update unit tests
- [x] manual behavior check

## Screenshots (before/after)

<details>
<summary>pools</summary>
<img width="927" alt="pools" src="https://user-images.githubusercontent.com/11639738/175181525-fc182dc1-a48b-4862-963f-c73f2cb97170.png">
</details>

<details>
<summary>pools list</summary>
<img width="927" alt="pools_list" src="https://user-images.githubusercontent.com/11639738/175181582-ff800754-d917-4e84-9d04-82712bd5723e.png">
</details>

<details>
<summary>pools get</summary>
<img width="927" alt="pools_get" src="https://user-images.githubusercontent.com/11639738/175181655-2f49f94f-937b-4a78-a390-4b6ce47deb87.png">
</details>

<details>
<summary>pools set</summary>
<img width="927" alt="pools_set" src="https://user-images.githubusercontent.com/11639738/175181708-01f51822-b11e-4c8d-8f2e-398eea0ed60f.png">
</details>

<details>
<summary>pools delete</summary>
<img width="927" alt="pools_delete" src="https://user-images.githubusercontent.com/11639738/175181748-cbe7da30-19e6-4642-9086-680addd6f22e.png">
</details>

<details>
<summary>pools import</summary>
<img width="927" alt="pools_import" src="https://user-images.githubusercontent.com/11639738/175181795-5cae0823-0b3d-4e74-a4d9-62cb76d30ad3.png">
</details>

<details>
<summary>pools export</summary>
<img width="927" alt="pools_export" src="https://user-images.githubusercontent.com/11639738/175181846-bf971689-deb8-4ac6-b7d1-6ae756d5813d.png">
</details>

hankehly commented on airflow/cli/commands/pool.py:
```diff
@@ -0,0 +1,162 @@
+@click.argument("description")
+@click_verbose
+@suppress_logs_and_warning_click_compatible
+def pool_set(name, slots, description, verbose):
```
I removed unused `output` argument from the `set` and `delete` commands.

blag replied:
Hm, I'm not sure how I feel about this. Yes, we should remove it, but previously it was simply ignored.

I'm leaning towards keeping it in, and throwing a warning that it will be removed in a future version.

WDYT?

hankehly commented on tests/cli/commands/test_pool.py:
```diff
@@ -0,0 +1,119 @@
+        assert "foo" in result.output
+
+    def test_pool_list_with_args(self):
+        result = self.runner.invoke(pool.pools, ["list", "--output", "json"])
```
Some of these tests are interdependent/overlapping. The scope of this PR is "converting to click cli" so I will leave refactoring for another time.

hankehly commented:
@blag Please review at your earliest convenience.

blag commented on airflow/cli/commands/pool.py:
```diff
@@ -0,0 +1,162 @@
+@click.argument("slots")
+@click.argument("description")
+@click_verbose
+@suppress_logs_and_warning_click_compatible
```
Any reason you removed the `@cli_utils.action_cli` decorator on `set`, `delete`, and `import`?

hankehly replied:
I mistook it as a decorator to register non-click CLI commands! I see that is not the case (adding it back in)

blag commented on airflow/cli/commands/pool.py:
```diff
@@ -0,0 +1,162 @@
+        }
+
+    """
+    if not os.path.exists(filepath):
+        raise SystemExit(f"Missing pools file {filepath}")
```
Click can handle this automatically, see here:
https://click.palletsprojects.com/en/8.1.x/arguments/#file-path-arguments

If we do switch over to this, we will need to update the tests to expect a nonzero exit and not expect for a `SystemExit` exception.

hankehly replied:
> Hm, I'm not sure how I feel about this. Yes, we should remove it, but previously it was simply ignored.
> I'm leaning towards keeping it in, and throwing a warning that it will be removed in a future version.
> WDYT?

I also like the idea of keeping the interface the same and adding a warning.

blag commented on airflow/cli/commands/pool.py:
```diff
@@ -0,0 +1,162 @@
+    """Creates new pool with a given name and slots"""
+    api_client = get_current_api_client()
+    api_client.create_pool(name=name, slots=slots, description=description)
+    print(f"Pool {name} created")
```
I'd like to try to route all printed output from commands through Rich's Console object, if possible:

```suggestion
    from rich.console import Console
    console = Console()
    ...
    console.print(f"Pool {name} created")
```

hankehly replied:
Missed that, thanks.

blag reviewed (changes requested):
Looks good, just a few more small improvements to make.

Also, I'm trying to get the module names to match the subcommand names, so renaming `airflow/cli/commands/pool.py` to `airflow/cli/commands/pools.py` (plural) would make that more consistent.

hankehly commented on airflow/cli/commands/pools.py:
```diff
@@ -114,20 +121,18 @@ def pool_import(filepath, verbose):
 
 @pools.command("export")
-@click.argument("filepath")
+@click.argument("filepath", type=click.Path())
```
Specifying `type` as `click.Path()` here does not change the behavior, but makes the intent clearer

hankehly commented:
@blag Thanks for the review. afedf2f makes the following changes:

1. renamed modules `pool` to `pools`
2. added unused `output` argument to set/delete functions (to match the non-click version) and added deprecation warning
3. replaced `print` statements with `AirflowConsole().print`
4. specified `click.Path` type for `filepath` arguments
5. added (mistakenly deleted) `@cli_utils.action_cli` decorator to `set`, `delete` and `import` functions

Regarding [your comment](https://github.com/astronomer/airflow/pull/1501#discussion_r904673292) about testing conditions...
> If we do switch over to this, we will need to update the tests to expect a nonzero exit and not expect for a SystemExit exception.

It looks like an invalid `click.Path` raises a [`BadParameter`](https://github.com/pallets/click/blob/a8910b382d37cce14adeb44a73aca1d4e87c2413/src/click/types.py#L860) exception, which is then caught internally and re-raised by [`sys.exit`](https://github.com/pallets/click/blob/0aec1168ac591e159baf6f61026d6ae322c53aaf/src/click/core.py#L1069-L1073), so the result is still a `SystemExit`. We can check for `BadParameter` by asserting `result.exit_code == 2` but this doesn't distinguish between parameters, so it's not much different than the current implementation. I'm going to leave it as-is. Let me know your thoughts.

blag commented:
Thank you for all of the work you've done for this.

From [this discussion](https://github.com/apache/airflow/pull/24591#pullrequestreview-1019268731), there is more groundwork and research necessary before we can evaluate rewriting the CLI with Click.
